### ISSUE TITLE
iOS- 1.0.2-51- Left menu shown in the center and need to reload App to fix it

### ISSUE THREAD

**Issue by mallenexpensify** [Expensify/Expensify.cash#2180]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
Job on Upwork https://www.upwork.com/jobs/~01acd19801f956e8ed

___

## Expected Result:
App should be open and show chat as usually.

## Actual Result:
App is opened but left menu is shown in the middle. Need to toggle device orientation again to fix it.

## Action Performed:
1. Launch App 
2. Open any chat 
3. Switch orientation of iPad from landscape to portrait
4. Switch orientation back to landscape
5. Tap on home button on iPAD to put app into the background
6. Open App from background

## Platform:
iOS - iPad 

**Version Number:**  1.0.2-51

**Notes/Photos/Videos:**  
![image](https://user-images.githubusercontent.com/22508304/113214573-8a37dd80-922e-11eb-8597-bc8f6299aa76.png)

**Expensify/Expensify Issue URL:**
https://github.com/Expensify/Expensify/issues/157516

**Comment by Amazing1974** [comment]
can you share the source code? then i can fix it.

**Comment by mallenexpensify** [comment]
Hi @Amazing1974 everything you should need to propose a solution should be in 
https://github.com/Expensify/Expensify.cash/blob/master/CONTRIBUTING.md
and/or 
https://github.com/Expensify/Expensify.cash/blob/master/README.md

**Comment by mallenexpensify** [comment]
@marcaaron assigned to you for incoming proposal reviews since you added the External label to the E/E issue

**Comment by marcaaron** [comment]
Gonna unassign myself from this one as the process changed on that. 
Adding the `Exported` label should auto-assign someone now.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mallenexpensify** [comment]
on hold to see if https://github.com/Expensify/Expensify.cash/issues/2051 magically fixes this too.

**Comment by deetergp** [comment]
🤞

**Comment by tugbadogan** [comment]
I debugged this issue and found that react native's `Dimensions` module is behaving weirdly on older iPad devices. When the app is sent to background, `onDimensionChange` callback function is called with wrong window dimensions before it's called with the right dimensions.

While using iPad in portrait mode (width: 768, height: 1024), when I send the app to the background this function is first called with (width: 1024, height: 768) and then with (width: 768, height: 1024). Components are rendered based on the first call and it's causing elements to be positioned badly on the screen.

I'm still trying to find out why it behaves like this on older iPad versions while it's working as expected on iPhone and iPad Pro.

**Comment by marcaaron** [comment]
Could be related to this ? https://github.com/facebook/react-native/issues/29290

**Comment by tugbadogan** [comment]
@marcaaron 
> Could be related to this ? [facebook/react-native#29290](https://github.com/facebook/react-native/issues/29290)

Yes, it looks similar. I couldn't find any workaround or temporary solution yet.

**Comment by deetergp** [comment]
Any word on your findings, @tugbadogan?

**Comment by tugbadogan** [comment]
Yes, I have a proposal for this issue now. 

Since this problem is caused by Dimensions native class is calling 'change' callback too many times with wrong dimensions, we can use `debounce` to slow down listening these changes, so we can avoid breaking call sites by triggering too many window dimension updates.

<img width="746" alt="Screenshot 2021-04-22 at 02 25 51" src="https://user-images.githubusercontent.com/3853003/115642007-19a64d00-a312-11eb-8d20-34332c971e47.png">

Making these changes resolve this issue.

**Comment by marcaaron** [comment]
Thanks, but maybe there's more to it? I applied this and was able to break it quickly by just changing the orientation a few times.

**Comment by tugbadogan** [comment]
Which device/emulator are you using to test this? I couldn't reproduce the issue after this fix even though I changed the orientation many times.

https://user-images.githubusercontent.com/3853003/115761826-ad236080-a39a-11eb-907f-ff9692ee7ba0.mov

**Comment by tugbadogan** [comment]
Hmm, I was able to reproduce broken view on orientation change on an iPhone (I couldn't find a reliable way to reproduce though). This looks like a separate issue then iPad having broken view when the app is sent to background. It looks like a bug on `Drawer.Navigator`.

**Comment by marcaaron** [comment]
>Hmm, I was able to reproduce broken view on orientation change on an iPhone

I'm confused as the current discussion is not about iPhone.

> This looks like a separate issue then iPad having broken view when the app is sent to background. It looks like a bug on Drawer.Navigator.

Not too sure what you are referring to exactly, but if you have discovered another issue please create a new issue or add context to an existing one.

**Comment by mallenexpensify** [comment]
I doubled price to $500 in Upwork https://www.upwork.com/jobs/~01e2db72cd3240af67

**Comment by mallenexpensify** [comment]
The link above wasn't getting any 👀 so I closed it and created another with the $500 price. 
https://www.upwork.com/jobs/~01acd19801f956e8ed

**Comment by ViniSchonardie** [comment]
Im having some difficulties to reproduce the error on Ipad Pro 9-7 emulator;
But a good option would be to transform the hook to functional component and use this react native hook 
import { useWindowDimensions } from 'react-native';
https://reactnative.dev/docs/usewindowdimensions
Its more reliable;

**Comment by ViniSchonardie** [comment]
Also, i am watching the React Native changelog, and aparently they made some changes about this on 0.63.4, and the Expensify.cash its on 0.63.3...

https://github.com/react-native-community/releases/blob/master/CHANGELOG.md
Dimension update events are now properly sent following orientation change (0e9296b95d by @ajpaulingalls)
https://github.com/facebook/react-native/commit/0e9296b95da06789121f052e6cd6d7cac808464c

**Comment by marcaaron** [comment]
>Its more reliable;

What makes it more reliable?

>Dimension update events are now properly sent following orientation change (0e9296b95d by @ajpaulingalls)

Thanks, but that commit looks like it changed Java code and we are experiencing an issue on iOS.

**Comment by ViniSchonardie** [comment]
@marcaaron well, even Dimensions documentation discourage you to use.
https://reactnative.dev/docs/dimensions

Dimensions Doc about it:
`useWindowDimensions is the preferred API for React components. Unlike Dimensions, it updates as the window's dimensions update. This works nicely with the React paradigm.`

Also for any of the cases, they advise you to always call the `Dimensions.get('window');` again, may be a dumb recheck seeing that you receive the newDimensions... `onDimensionChange(newDimensions) {` but may help...

Dimensions doc about it:
Although dimensions are available immediately, they may change (e.g due to device rotation, foldable devices etc) so any rendering logic or styles that depend on these constants should try to call this function on every render, rather than caching the value (for example, using inline styles rather than setting a value in a StyleSheet).

**Comment by marcaaron** [comment]
`useWindowDimensions` is quite similar to the HOC we have created and itself relies on the `Dimensions` so the docs may simply be trying to discourage people from using `Dimensions` without any sort of listener?

See the `react-native` source [here](https://github.com/facebook/react-native/blob/dc80b2dcb52fadec6a573a9dd1824393f8c29fdc/Libraries/Utilities/useWindowDimensions.js#L15-L38)

And our own [here](https://github.com/Expensify/Expensify.cash/blob/main/src/components/withWindowDimensions.js)

>Also for any of the cases, they advise you to always call the Dimensions.get('window'); again, may be a dumb recheck seeing that you receive the newDimensions... onDimensionChange(newDimensions) { but may help...

Sorry, don't quite understand what you mean here.

**Comment by ViniSchonardie** [comment]
What i was saying was to make the get at every change besided getting the newDimensions as a parameter...
Equal to what the hook does...
handleChange({window: Dimensions.get('window')});

**Comment by marcaaron** [comment]
Ah, OK I finally understand now! Apologies, I missed what was happening in the hook.

So, I don't think it would be terrible to update things to use the hook? But after some quick testing the problem persists after the suggested change.

It is consistently reproducible if we:

1. Start in portrait
2. Switch to landscape
3. Background the app
4. Open again
5. 💥 

https://user-images.githubusercontent.com/32969087/117091202-3c633900-acf6-11eb-8ebc-6a8ed2af78b3.mp4

**Comment by marcaaron** [comment]
I confirmed this [proposal from @tugbadogan](https://github.com/Expensify/Expensify.cash/issues/2180#issuecomment-824475188) does fix the issue. So perhaps that is a clue about how it can happen.

**Comment by mallenexpensify** [comment]
@marcaaron should we hire @tugbadogan in Upwork for this job?

**Comment by marcaaron** [comment]
@mallenexpensify I'm not sure. The solution is a workaround and seems to point to a greater issue in `react-native`. There is still this open issue -> https://github.com/facebook/react-native/issues/29290

We could implement that solution or try to fix it upstream. Anyways, I'll let @deetergp decide what to do from here.

**Comment by deetergp** [comment]
It feels kinda hacky, but if the problem is with `react-native` core, we can either wait for them to fix it or go with the hack. I think I'm inclined to go with @tugbadogan's workaround.

**Comment by marcaaron** [comment]
I'm cool with that as well. What about if we do that but also create a follow up issue to address it later? In this case, it might mean that we just revisit things when  it gets fixed upstream in `react-native` so we aren't sitting around waiting for the fix.

**Comment by marcaaron** [comment]
Let's do the fix for now and create an entirely separate issue to track the progress of [this issue](https://github.com/facebook/react-native/issues/29290) ?

**Comment by mallenexpensify** [comment]
I needed to create a new Upwork job (https://www.upwork.com/jobs/~012e79714137f7faa6), the previous one expired. @tugbadogan I think I invited you in Upwork, can you please confirm here in this GH you're still interested in fixing this **and** also apply in Upwork?   Thanks

**Comment by tugbadogan** [comment]
Thanks for the invitation @mallenexpensify , I've now applied via Upwork.

**Comment by mallenexpensify** [comment]
@tugbadogan two Upwork issues somehow got created for this same job, I accidentally 'hired' you for the duplicate before finding both. I then canceled one and requested a refund for it, can you please OK the refund for the duplicate issue?  `The freelancer will receive a request to approve the refund of your $500.00 deposit.` 
Sorry for the mix up, still not sure exactly what happened

**Comment by mallenexpensify** [comment]
Hi @tugbadogan we should be all set, you're hired for this GH via this Upwork job
https://www.upwork.com/jobs/~012e79714137f7faa6

Let me know if anything doesn't look right.

**Comment by mallenexpensify** [comment]
Apolgies @tugbadogan this is all my fault and is still confusing me. 
Can you see this job? https://www.upwork.com/jobs/~01acd19801f956e8ed
There should be a pending 'hire' invitation for you there. Accept that one and we should be good to go (I really hope so)

**Comment by mallenexpensify** [comment]
Will pay on 5/19 if there are no regressions

**Comment by isagoico** [comment]
1 tester was able to reproduce this following these steps: 

1. Open e.cash on iPad 
2. Change the view from landscape to portrait
2. Send the app to background and open another app (can be any) 
3. Immediately navigate to the e.cash app. 

Video here: 

https://user-images.githubusercontent.com/44479856/119015039-8a1ea900-b966-11eb-8d29-4cba4431fb26.mp4

**Comment by parasharrajat** [comment]
@tugbadogan Are you trying to update to v6? Just curious.

**Comment by tugbadogan** [comment]
> @tugbadogan Are you trying to update to v6? Just curious.

Yes

**Comment by tugbadogan** [comment]
I upgraded react navigation to 6.x, react native reanimated to 2.1.0 and tested this issue and https://github.com/Expensify/Expensify.cash/issues/2447. It looks like these issues are resolved in the new version.

It has a couple of breaking changes though that we need to fix to make sure it's not causing regressions to other features, e.g. `dangerouslyGetParent()` function is removed.

@parasharrajat Is there anyone else working on upgrading the react navigation version or another issue blocked by this?

**Comment by parasharrajat** [comment]
@tugbadogan No. But I believe that it would resolve a couple of issues.

**Comment by marcaaron** [comment]
I'd be curious to see an upgrade PR. I think v6 is still in a beta maybe?

**Comment by parasharrajat** [comment]
@tugbadogan I think you can start a PR and mention all the issues there so that they can be tracked. I am sure others will be happy to help here.

**Comment by mallenexpensify** [comment]
Reopening since @isagoico was able to reproduce the issue https://github.com/Expensify/Expensify.cash/issues/2180#issuecomment-845272513

I'm unsure of the best path forward, should we put a HOLD on this until we have react native updated, to see if that will fix the issue?

**Comment by mallenexpensify** [comment]
@Jag96 Should I plan to pay @tugbadogan on June 15th for this?

**Comment by Jag96** [comment]
@mallenexpensify yes, updated the title 👍

**Comment by tugbadogan** [comment]
@mallenexpensify It looks like Upwork approved the milestone and paid me automatically after some time passed.

**Comment by mallenexpensify** [comment]
Hi @tugbadogan is this issue the 'exact' same as this one? https://github.com/Expensify/Expensify.cash/issues/2447

The combo of these two has been confusing me for WEEKS

**Comment by tugbadogan** [comment]
@mallenexpensify This issue and #2447 are both caused by some animation bugs in react-navigation library. These issues are separate symptoms of similar bugs in react-navigation.

* In this issue, react navigation fails to update the position of the left menu when orientation is changed to landscape, so it's displayed in the center and UI breaks.
* In #2447, when device orientation is changed to portrait, react-navigation fails to move the left menu out of screen, left menu is displayed correctly on the screen, but chat buttons are not clickable. 

By upgrading react-navigation to 6.x, we started using React Animated 2 which I expect to solve positioning issues, so both of these issues can be closed.

**Comment by parasharrajat** [comment]
> By upgrading react-navigation to 6.x, we started using React Animated 2
@tugbadogan I believe this perception is wrong here as per the docs https://reactnavigation.org/docs/6.x/drawer-navigator#uselegacyimplementation. And you have not set the flag explicitly.

**Comment by tugbadogan** [comment]
The documentation says 

```
This defaults to true in following cases:
* Reanimated 2 is not configured
* App is connected to Chrome debugger (Reanimated 2 cannot be used with Chrome debugger)
* App is running on Web

Otherwise, it defaults to false
```

That prop defaults to `false` when Reanimated 2 is configured. This issue and other issue come back when I set `useLegacyImplementation` prop to `true`.

**Comment by parasharrajat** [comment]
I meant that we should set `useLegacyImplementation` as `false` explicitly as I think it says for any of those cases it's true.

**Comment by tugbadogan** [comment]
> I meant that we should set `useLegacyImplementation` as `false` explicitly as I think it says for any of those cases it's true. But its fine if issue is fixed.

Yeah, it'd be better to set it false to enable Reanimated 2 on web and have consistency. That's need to be tested thoroughly on web though.

**Comment by tugbadogan** [comment]
I tried and it throws some exceptions, it may not be ready for react-native-web yet or we need to fix a couple of more things to enable it on web.

<img width="491" alt="Screenshot 2021-06-16 at 00 32 27" src="https://user-images.githubusercontent.com/3853003/122136419-5f772180-ce3a-11eb-905c-b414d8aa568e.png">

**Comment by parasharrajat** [comment]
@tugbadogan 
I think Reanimated 2 is not configured on Android as well. I see a warning 
![image](https://user-images.githubusercontent.com/24370807/122138204-dd035780-ce63-11eb-9ed7-c8483b83e80e.png)

**Comment by parasharrajat** [comment]
We may be missing these https://docs.swmansion.com/react-native-reanimated/docs/2.1.0/installation#android

**Comment by mallenexpensify** [comment]
This should have been paid via Upwork

**Comment by tugbadogan** [comment]
Hi @mallenexpensify, 

Other issue (https://github.com/Expensify/Expensify.cash/issues/2447) is closed now. I think we can close this issue as well.

**Comment by mallenexpensify** [comment]
Let's do it @tugbadogan !!!

**Comment by dklymenk** [comment]
Hello, I think the job posting for this issue is no longer relevant, right? https://www.upwork.com/jobs/~012e79714137f7faa6

**Comment by mallenexpensify** [comment]
Closed, thanks @dklymenk .  This issue was a nightmare for me, think it got double posted months ago.
